### PR TITLE
Fix a genus function and allow parity variations...

### PR DESCRIPTION
**Pull request by StevellM**
...and also relax the signatures in the constructor `torsion_quadratic_module`

### CODE REVIEW

**Comment by thofma** [comment]
@simonbrandhorst can you have a look? this is above my pay grade

File: src/QuadForm/Torsion.jl [hunk 1]
```diff
@@ -70,22 +74,24 @@ function torsion_quadratic_module(M::ZZLat, N::ZZLat; gens::Union{Nothing, Vecto
...
+  if gens !== nothing  && length(gens) > 0
     gens_lift = gens
   else
     gens_lift = Vector{QQFieldElem}[reshape(collect(change_base_ring(FlintQQ, mS(s).coeff) * BM), :) for s in Hecke.gens(S)]
   end
 
   num = basis_matrix(M) * gram_matrix(ambient_space(M)) * transpose(basis_matrix(N))
   if iszero(modulus)
-    modulus = reduce(gcd, num; init = zero(QQFieldElem))
+    _modulus = reduce(gcd, num; init = zero(QQFieldElem))
```
Comment (simonbrandhorst) [inline src/QuadForm/Torsion.jl:85]: why all these underscores?

File: src/QuadForm/Torsion.jl [hunk 2]
```diff
@@ -1695,16 +1701,29 @@ function brown_invariant(T::TorQuadModule)
   return brown
 end
 
+function _as_finite_bilinear_module(T::TorQuadModule)
+  n = modulus_bilinear_form(T)
+  return torsion_quadratic_module(cover(T), relations(T); modulus = n, modulus_qf = n)
```
Comment (simonbrandhorst) [inline src/QuadForm/Torsion.jl:1706]:
```suggestion
  n = modulus_bilinear_form(T)
  if n == modulus_quadratic_form(T)
    return T
  end
  return torsion_quadratic_module(cover(T), relations(T); modulus = n, modulus_qf = n)
```

File: src/QuadForm/Torsion.jl [hunk 3]
```diff
@@ -1695,16 +1701,29 @@ function brown_invariant(T::TorQuadModule)
...
+end
+
 @doc raw"""
-    genus(T::TorQuadModule, signature_pair::Tuple{Int, Int}) -> ZZGenus
+    genus(T::TorQuadModule, signature_pair::Tuple{Int, Int};
+                            parity::RationalUnion = modulus_quadratic_form(T))
```
Comment (simonbrandhorst) [inline src/QuadForm/Torsion.jl:1711]: The parity must not be a rational number. Hence `RationalUnion` is missleading.

File: src/QuadForm/Torsion.jl [hunk 4]
```diff
@@ -1695,16 +1701,29 @@ function brown_invariant(T::TorQuadModule)
...
+If no such genus exists, raise an error.
+
+By default the `parity` is taken as the parity of the quadratic form on `T`.
 
 # Reference
 [Nik79](@cite) Corollary 1.9.4 and 1.16.3.
 """
-function genus(T::TorQuadModule, signature_pair::Tuple{Int, Int})
+function genus(T::TorQuadModule, signature_pair::Tuple{Int, Int}; parity::RationalUnion = modulus_quadratic_form(T))
+  @req modulus_bilinear_form(T) == 1 "Modulus for the bilinear form should be 1"
```
Comment (simonbrandhorst) [inline src/QuadForm/Torsion.jl:1725]:
```suggestion
  @req modulus_bilinear_form(T) == 1 "Modulus for the bilinear form must be 1"
```

File: src/QuadForm/Torsion.jl [hunk 5]
```diff
@@ -1695,16 +1701,29 @@ function brown_invariant(T::TorQuadModule)
...
+  @req modulus_quadratic_form(T) == 1 || modulus_quadratic_form(T) == 2 "Modulus for the quadratic form should be 2"
```
Comment (simonbrandhorst) [inline src/QuadForm/Torsion.jl:1726]:
```suggestion
  @req modulus_quadratic_form(T) == 1 || modulus_quadratic_form(T) == 2 "Modulus for the quadratic form must be 2"
```

File: src/QuadForm/Torsion.jl [hunk 6]
```diff
@@ -1751,6 +1771,11 @@ function genus(T::TorQuadModule, signature_pair::Tuple{Int, Int})
   end
   # This genus has the right discriminant group
   # but it may be empty
+  #
+  # in this case the blocks of scales 1, 2, 4 are under determined
```
Comment (simonbrandhorst) [inline src/QuadForm/Torsion.jl:1775]: This part of the comment still belongs to the odd case where modulus_quadratic_form(T) == 1.
The following two lines can stay.

File: src/QuadForm/Torsion.jl [hunk 7]
```diff
@@ -1761,13 +1786,10 @@ function genus(T::TorQuadModule, signature_pair::Tuple{Int, Int})
...
   end
-  if modulus_quadratic_form(T) == 1
-    # in this case the blocks of scales 1, 2, 4 are under determined
-    # make sure the first 3 symbols are of scales 1, 2, 4
-    # i.e. their valuations are 0, 1, 2
-
-    # the form is odd
-    block0 = [b for b in _blocks(sym2[1]) if b[4] == 1]
+  if modulus_quadratic_form(T) == 1 || parity == 1
+    _o = mod(parity, 2)
```
Comment (simonbrandhorst) [inline src/QuadForm/Torsion.jl:1790]: ???

File: src/QuadForm/Torsion.jl [hunk 8]
```diff
@@ -1831,15 +1853,19 @@ function genus(T::TorQuadModule, signature_pair::Tuple{Int, Int})
...
-    is_genus(T::TorQuadModule, signature_pair::Tuple{Int, Int}) -> Bool
+    is_genus(T::TorQuadModule, signature_pair::Tuple{Int, Int};
+                               parity::RationalUnion = modulus_quadratic_form(T)) -> Bool
+
+Return if there is an integral lattice whose discriminant form has the bilinear
+form of `T`, whose signatures match `signature_pair` and which is of parity
+`parity`.
 
-Return if there is an integral lattice with this signature and discriminant form.
+By deautl the `parity` is the parity of the quadratic form on `T`.
```
Comment (simonbrandhorst) [inline src/QuadForm/Torsion.jl:1864]:
```suggestion
By default the `parity` is the parity of the quadratic form on `T`.
```

**Review by simonbrandhorst — CHANGES_REQUESTED** [review]

File: src/QuadForm/Torsion.jl (hunk 1 above)
Comment (thofma) [reply src/QuadForm/Torsion.jl:85]: If the input is `1//1`, then `modulus` would change it's type here.

File: src/QuadForm/Torsion.jl (hunk 3 above)
Comment (StevellM) [reply src/QuadForm/Torsion.jl:1711]: Yes, it has to be a rational number. In any case, it should be restricted to 1 or 2 for us, but then if we enter a `Int`, `ZZRingElem` or `QQFieldElem`, then everything fits in `RationalUnion`.

File: src/QuadForm/Torsion.jl (hunk 7 above)
Comment (StevellM) [reply src/QuadForm/Torsion.jl:1790]: If `parity == 1`, I want 1 for the next line, and if `parity == 2`, I want 0.

**Review by simonbrandhorst — APPROVED** [review]

**Pull request merged**